pkg/cache/v3/linear.go

```diff
@@ -312,90 +370,72 @@ func (cache *LinearCache) GetResources() map[string]types.Resource {
 	return resources
 }
 
-func (cache *LinearCache) CreateWatch(request *Request, _ Subscription, value chan Response) (func(), error) {
+func (cache *LinearCache) CreateWatch(request *Request, sub Subscription, value chan Response) (func(), error) {
 	if request.GetTypeUrl() != cache.typeURL {
 		return nil, fmt.Errorf("request type %s does not match cache type %s", request.GetTypeUrl(), cache.typeURL)
 	}
-	// If the version is not up to date, check whether any requested resource has
-	// been updated between the last version and the current version. This avoids the problem
-	// of sending empty updates whenever an irrelevant resource changes.
-	stale := false
-	staleResources := []string{} // empty means all
-
-	// strip version prefix if it is present
-	var lastVersion uint64
-	var err error
-	if strings.HasPrefix(request.GetVersionInfo(), cache.versionPrefix) {
-		lastVersion, err = strconv.ParseUint(request.GetVersionInfo()[len(cache.versionPrefix):], 0, 64)
-	} else {
-		err = errors.New("mis-matched version prefix")
-	}
 
-	watch := ResponseWatch{Request: request, Response: value}
+	// A major difference between delta and sotw is the ability to not resend everything when connecting to a new control-plane
+	// In delta the request provides the version of the resources it does know, even if the request is wildcard or does request more resources
+	// In sotw the request only provides the global version of the control-plane, and there is no way for the control-plane to know if resources have
+	// been added since in the requested resources. In the context of generalized wildcard, even wildcard could be new, and taking the assumption
+	// that wildcard implies that the client already knows all resources at the given version is no longer true.
+	// We could optimize the reconnection case here if:
+	//  - we take the assumption that clients will not start requesting wildcard while providing a version. We could then ignore requests providing the resources.
+	//  - we use the version as some form of hash or resources known, and we can then consider it as a way to correctly verify whether all resources are unchanged.
+	// For now it is not done as:
+	//  - for the first case, while the protocol documentation does not explicitly mention the case, it does not mark it impossible and explicitly references unsubscribing from wildcard
+	//  - for the second one we could likely do it with little difficulty if need be
```

Comment:
Yes. We could, for example, have the version be a string representation of a map of resource names -> hash, with a special value meaning meaning that the resource did not exist, and perhaps special case for wildcards. We'll see how much network bandwidth we waste with reconnections.

Reply:
Yes as discussed, I don't want to add such additional change in this PR. Can be followed up later

Reply:
Makes sense, let's see if we need that in the future.